### PR TITLE
Space supporter process write permissions

### PR DESCRIPTION
**Pull request by belinda-liu**
Hello CAPI team,

This is a PR to allow the `space supporter` role to access POST, PATCH, and DELETE actions on process endpoints. This addresses [this issue](https://github.com/cloudfoundry/cloud_controller_ng/issues/2211) and is part of the larger track of work to add the `space supporter` role to CCNG.
Please let us know if you have any feedback.

Thanks,
@galenhammond && @belinda-liu 

Thanks for contributing to cloud_controller_ng. To speed up the process of reviewing your pull request please provide us with:

* [x] I have reviewed the [contributing guide](https://github.com/cloudfoundry/cloud_controller_ng/blob/main/CONTRIBUTING.md)

* [x] I have viewed, signed, and submitted the Contributor License Agreement

* [x] I have made this pull request to the `main` branch

* [x] I have run all the unit tests using `bundle exec rake`

* [ ] I have run [CF Acceptance Tests](https://github.com/cloudfoundry/cloud_controller_ng#cf-acceptance-tests-cats)

### CODE REVIEW

File: docs/v3/source/includes/resources/processes/_list_for_app.md.erb [hunk 1]
```diff
@@ -40,12 +40,13 @@ Name | Type | Description
...
- |
+Roles | Notes |
 --- | ---
 Admin |
 Admin Read-Only |
 Global Auditor |
 Org Manager |
 Space Auditor |
 Space Developer |
+Space Application Developer | Experimental
```
Comment (sweinstein22) [inline docs/v3/source/includes/resources/processes/_list_for_app.md.erb:51]: I haven't looked through this PR extensively yet, but just at first glance I believe this should actually be `Space Application Supporter` at this stage, and we've been aiming to alphabetize the roles. Let me know if I missed something there

**Review by sweinstein22 — COMMENTED** [review]
Pretty sure this one also covers a previously opened PR, https://github.com/cloudfoundry/cloud_controller_ng/pull/2331, as the commit from that PR is also listed here. Going to do acceptance on the full list of endpoints from the associated endpoint here.

## Acceptance

### Behavior Check

#### `GET` endpoints

Ran `cf curl` as both a space developer and a space application supporter against the following endpoints, checked the output manually and confirmed that diffs only included redacted info, slight variations in memory or cpu usage, timestamps, and uptime. Confirmed that redacted start command for apps was deliberate decision.

Endpoints:
`/v3/processes`, `/v3/processes/:guid`, `/v3/processes/:guid/stats`, `/v3/apps/:guid/processes`, `/v3/apps/:guid/processes/:type`, `/v3/apps/:guid/processes/:type/stats`

```
$ diff space-app-supporter-processes space-dev-processes

$ diff space-app-supporter-process-guid space-dev-process-guid
6c6
<   "command": "[PRIVATE DATA HIDDEN]",
---
>   "command": "bundle exec rackup config.ru -p $PORT",

$ diff space-app-supporter-processes-guid-stats space-dev-processes-guid-stats
8c8
<       "uptime": 590724,
---
>       "uptime": 590258,
29,31c29,31
<         "time": "2021-06-29T19:36:19+00:00",
<         "cpu": 0.005408429952503851,
<         "mem": 41859299,
---
>         "time": "2021-06-29T19:28:33+00:00",
>         "cpu": 0.0052539787788561994,
>         "mem": 42099598,
40c40
<       "uptime": 82423,
---
>       "uptime": 81957,
61,63c61,63
<         "time": "2021-06-29T19:36:19+00:00",
<         "cpu": 0.005274389094794207,
<         "mem": 52461568,
---
>         "time": "2021-06-29T19:28:33+00:00",
>         "cpu": 0.005232974464376784,
>         "mem": 52341418,
$ diff space-app-supporter-app-processes space-dev-app-processes

$ diff space-app-supporter-app-processes-web space-dev-app-processes-web
6c6
<   "command": "[PRIVATE DATA HIDDEN]",
---
>   "command": "bundle exec rackup config.ru -p $PORT",
$ diff space-app-supporter-app-processes-web-stats space-dev-app-processes-web-stats
8c8
<       "uptime": 914076,
---
>       "uptime": 913995,
18c18
<         "time": "2021-06-29T19:36:53+00:00",
---
>         "time": "2021-06-29T19:35:31+00:00",
```

#### `PATCH` endpoints
`cf curl -X PATCH /v3/processes/12ff2e1e-474d-4792-abc0-ce92ca7a4db3 -d '{"command": "rackup all the things", "metadata": {"labels": {"key": "value2"}, "annotations": {"note": "less detailed information"}}}'` as space app supporter
`cf curl -X PATCH /v3/processes/12ff2e1e-474d-4792-abc0-ce92ca7a4db3 -d '{"command": "rackup","metadata": {"labels": {"key": "value"},"annotations": {"note": "detailed information"}}}'` as space developer. Checked output manually and confirmed that diff was only in values provided and update timestamps:

```
$ diff space-app-supporter-patch-process space-dev-patch-process
4c4
<   "updated_at": "2021-06-29T20:36:50Z",
---
>   "updated_at": "2021-06-29T20:37:53Z",
6c6
<   "command": "rackup all the things",
---
>   "command": "rackup",
31c31
<       "key": "value2"
---
>       "key": "value"
34c34
<       "note": "less detailed information"
---
>       "note": "detailed information"
```

Food for thought, it feels a little weird that the call to `PATCH` is accessible, including allowing space app supporters to change the app `command`, but that `command` gets redacted in the return body of `GET /v3/processes/:guid`.

#### `POST` endpoints

Ran `cf curl -X POST` against `/v3/processes/:guid/actions/scale` and `/v3/apps/:guid/processes/:type/actions/scale` as both a space dev and space application supporter. Differed numbers on each call to ensure they were taking effect, confirmed that output looked correct manually, and checked to make sure only differences in output were variable numbers and updated at timestamps:

```
$ diff space-app-supporter-app-web-process-scale space-dev-app-web-process-scale
4c4
<   "updated_at": "2021-06-29T20:46:15Z",
---
>   "updated_at": "2021-06-29T20:47:48Z",
7,9c7,9
<   "instances": 3,
<   "memory_in_mb": 256,
<   "disk_in_mb": 1024,
---
>   "instances": 6,
>   "memory_in_mb": 1024,
>   "disk_in_mb": 2048,

$ diff space-app-supporter-process-scale space-dev-process-scale
4c4
<   "updated_at": "2021-06-29T20:46:24Z",
---
>   "updated_at": "2021-06-29T20:46:59Z",
7,9c7,9
<   "instances": 5,
<   "memory_in_mb": 256,
<   "disk_in_mb": 1024,
---
>   "instances": 4,
>   "memory_in_mb": 1024,
>   "disk_in_mb": 256,
```

#### `DELETE` endpoints

Ran `cf curl -X DELETE` against `/v3/processes/:guid/instances/:index` and `/v3/apps/:guid/processes/:type/instances/:index` as space application supporter. Ran `watch "cf curl /v3/apps/3b0ece13-bce9-4d59-8cd6-5151d9b36ec6/processes/web/stats | grep state"` simultaneously to ensure app instances actually were terminated. 

Output is an empty object for both space app supporters and space developers:
```
$ cf curl -X DELETE /v3/processes/3b0ece13-bce9-4d59-8cd6-5151d9b36ec6/instances/1
{}
$ cf curl -X DELETE /v3/apps/3b0ece13-bce9-4d59-8cd6-5151d9b36ec6/processes/web/instances/0
{}
```

**Review by sweinstein22 — CHANGES_REQUESTED** [review]
### Code Check

I'm generally hesitant to have contexts that are role specific, e.g. `when the user is assigned the space_supporter role`. In other PRs from community members its resulted in folks grabbing a whole context block and duplicating it, just changing the role, which leaves us with really hefty request specs.

It seems like the intent is to check the content of the response body and the behavior, would it be reasonable to change the context to indicate that it's meant to be a content and behavioral check (not mentioning the roles) and just leave the roles permissions to a shared example context?

### Docs Check
Looks like we missed some endpoints in the docs:
- `PATCH /v3/processes/:guid`
- `POST /v3/processes/:guid/actions/scale` and `POST /v3/apps/:guid/processes/:type/actions/scale`
- `DELETE /v3/processes/:guid/instances/:index` and `DELETE /v3/apps/:guid/processes/:type/instances/:index`

There's also a docs change that I've left a comment on, which I believe needs to be addressed. (for the endpoint `GET /v3/apps/:guid/processes`)

### Behavioral Question

@monamohebbi thoughts on the Space Application Supporter being able to PATCH a process's `command` but not being able to see it when curling `GET /v3/processes/:guid`?

**Comment by galenhammond** [comment]
Hey Sarah, thanks for the feedback! We've pushed a commit updating the docs.

In terms of the role specific context, we actually intentionally wanted to call out that the two cases are for the space supporter role. We were unable to use the shared examples for permission checks on the process scale endpoints, so we added these two one-off examples to complete our test coverage for the space supporter role. We think that changing the title of these contexts could potentially be misleading, but we're curious to hear your input!

Best,
@belinda-liu & Galen

**Comment by belinda-liu** [comment]
We had a quick discussion with @sweinstein22 about the role-specific tests. She pointed out that the tests may be using the wrong headers, which turned out to be true. I just pushed a commit to use the correct headers, with some slight formatting to hopefully make the test more explicit that it's using the `space supporter` role.

**Review by sweinstein22 — APPROVED** [review]
Looks great, thanks for the PR!

**Pull request merged**